### PR TITLE
Bug 1923689 - Implement basic configuration processing on SearchEngineSelector.

### PR DESCRIPTION
**Pull request by Standard8**
This implements processing the main parts of the search configuration JSON into the Rust structure using serde_json. Currently the processing applies to:

* The identifier and base properties of the engine records.
* The default engine records.

As variant and environment handling is not yet implemented, the filter function will return all engines defined in the configuration.

These functions are not currently used, hence no breaking API changes, similarly with change log.

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [X] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [X] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [X] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [X] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [X] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due diligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/search/src/filter.rs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::{
+    error::Error, JSONEngineBase, JSONEngineRecord, JSONEngineUrl, JSONEngineUrls,
+    JSONSearchConfigurationRecords, RefinedSearchConfig, SearchEngineDefinition, SearchEngineUrl,
+    SearchEngineUrls, SearchUserEnvironment,
+};
+
+impl SearchEngineUrl {
+    pub(crate) fn from_engine_url(url: &JSONEngineUrl) -> Self {
```
Comment (Standard8) [inline components/search/src/filter.rs:12]: I'm not sure if this is the best way of handling these conversions. There appears to be multiple ways, and this seemed to make sense?

File: components/search/src/types.rs [hunk 2]
```diff
@@ -165,8 +174,8 @@ pub struct SearchEngineDefinition {
     pub name: String,
 
     /// The partner code for the engine. This will be inserted into parameters
-    /// which include `{partnerCode}`.
-    pub partner_code: Option<String>,
+    /// which include `{partnerCode}`. May be the empty string.
+    pub partner_code: String,
```
Comment (Standard8) [inline components/search/src/types.rs:178]: I made this change, as I think we'd generally replace `{partnerCode}` with the string, whether it is empty or not.

**Review by linabutler — CHANGES_REQUESTED** [review]
This looks wonderful, @Standard8; it's great to see this coming together!

I left a lot of comments—some are style nits; some are more substantive thoughts on the implementation—but they're all meant to start conversations, not a prescriptive "let's do it this way".

I'm happy to answer more questions, and drop links to the Rust book and examples if you'd like; and also, please feel free to push back on any suggestions that don't make sense.

I'm also really excited that we're adding another component to our collection. Rust! Rust! Rust! 🦀

File: components/search/src/configuration_types.rs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::{SearchEngineClassification, SearchUrlParam};
+use serde::Deserialize;
+
+/// This module defines the structures that we use for serde_json to parse
+/// the search configuration.
+
+/// The list of possible submission methods for search engine urls.
+#[derive(Debug, uniffi::Enum, PartialEq, Deserialize, Clone)]
```
[the remark this answers is not in the document]
Comment (Standard8) [reply components/search/src/configuration_types.rs:12]: I don't think I need this, it seems to work without that.

**Comment by Standard8** [comment]
@linabutler Thank you for all the excellent comments. I definitely like the idea of cloning up-front. Although it feels like cloning a large amount when we might use just a small amount, I agree it is better that way, so that we can reduce the individual allocations & make the code simpler.

File: components/search/src/configuration_types.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+pub(crate) struct JSONEngineOrdersRecord {
+    // TODO: Implementation.
+}
+
+/// Represents an individual record in the raw search configuration.
+#[derive(Debug, Deserialize)]
+#[serde(tag = "recordType", rename_all = "camelCase")]
+pub(crate) enum JSONSearchConfigurationRecords {
+    DefaultEngines(JSONDefaultEnginesRecord),
+    Engine(Box<JSONEngineRecord>),
```
[the remark this answers is not in the document]
Comment (linabutler) [reply components/search/src/configuration_types.rs:122]: Oh, I see! Nope, this is perfect; thanks! I didn't realize that the difference in sizes of the variants was so large.

File: components/search/src/filter.rs [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+            }
+            JSONSearchConfigurationRecords::Unknown => {
+                // Prevents panics if a new record type is added in future.
+            }
+        }
+    }
+
+    Ok(RefinedSearchConfig {
+        engines,
+        app_default_engine_id: default_engine_id.unwrap(),
```
Comment (linabutler) [inline components/search/src/filter.rs:80]: Is this `.unwrap()` temporary? If not, I'm wondering if we should think about returning an error here instead—`.unwrap()` will immediately `abort()` the main process on Desktop.

File: components/search/src/filter.rs [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+    for record in configuration {
+        match record {
+            JSONSearchConfigurationRecords::Engine(engine) => {
+                let result = extract_engine_config(&user_environment, engine);
+                engines.extend(result);
+            }
+            JSONSearchConfigurationRecords::DefaultEngines(default_engines) => {
+                default_engine_id = Some(default_engines.global_default);
+                default_private_engine_id.clone_from(&default_engines.global_default_private);
```
Comment (linabutler) [inline components/search/src/filter.rs:67]: Oooh, I haven't seen a use of `.clone_from()` in the wild before! Did Clippy recommend this?

File: components/search/src/selector.rs [hunk 7]
```diff
@@ -2,20 +2,31 @@
...
 #[derive(Default, uniffi::Object)]
-pub struct SearchEngineSelector {}
+pub struct SearchEngineSelector(Mutex<SearchEngineSelectorInner>);
 
 #[uniffi::export]
 impl SearchEngineSelector {
     #[uniffi::constructor]
     pub fn new() -> Self {
-        Self::default()
+        Self(Mutex::new(SearchEngineSelectorInner::default()))
```
Comment (linabutler) [inline components/search/src/selector.rs:29]:
```suggestion
        Self(Mutex::default())
```

(`Mutex<T>` implements `Default` if `T` implements `Default`—but your code looks good as-is, too, if you'd like to keep it explicit!)

**Review by linabutler — APPROVED** [review]
This looks great to me, thank you so much for folding in all my feedback! :shipit:

File: components/search/src/filter.rs (hunk 5 above)
Comment (Standard8) [reply components/search/src/filter.rs:80]: Yeah, it is temporary. I'm actually working on it in a patch for full handling of default engines, and changing this to be an `Option<String>` - in the worst case that we can't determine a default engine, we'll let the app decide how to handle it. As that's in a patch already, I won't change that here.

**Pull request merged**